panoptes_aggregation/tests/batch_aggregation/test_batch_aggregation.py

```diff
@@ -0,0 +1,124 @@
+    def test_run_aggregation(self, mock_aggregator, mock_wf_ext_conf):
+        mock_aggregator.process_wf_export.return_value = MagicMock()
+        mock_aggregator.process_cls_export.return_value = MagicMock()
+
+        mock_df = MagicMock()
+        test_extracts = {'question_extractor': mock_df}
+        batch_utils.batch_extract = MagicMock(return_value=test_extracts)
+        mock_reducer = MagicMock()
+        batch_utils.batch_reduce = mock_reducer
+
+        run_aggregation(1, 10, 100)
+        mock_aggregator.assert_called_once_with(1, 10, 100)
+        mock_wf_ext_conf.assert_called_once()
+        batch_utils.batch_extract.assert_called_once()
+        mock_df.to_csv.assert_called()
+        batch_utils.batch_reduce.assert_called()
+        self.assertEqual(mock_reducer.call_count, 2)
+
+        # The reducer's call list includes subsequent calls to to_csv, but the args are methods called on the mock
+        # rather than use the set values i.e. "<MagicMock name='BatchAggregator().output_path' id='140281634764400'>"
+        # mock_aggregator.workflow_id = '10'
+        # mock_aggregator.output_path = 'tmp/10'
+        # mock_reducer.assert_has_calls([
+        #         call(mock_df, {'reducer_config': {'question_reducer': {}}}),
+        #         call().to_csv('tmp/10/10_reducers.csv', mode='a'),
+        #         call(mock_df, {'reducer_config': {'question_consensus_reducer': {}}}),
+        #         call().to_csv('tmp/10/10_reducers.csv', mode='a'),
+        #     ])
+
+        # How do I test the specific instance of BatchAggregator rather than the mocked class?
+        # mock_aggregator.upload_files.assert_called_once()
+
+    @patch("panoptes_aggregation.batch_aggregation.os.mkdir")
+    @patch("panoptes_aggregation.batch_aggregation.Workflow")
+    @patch("panoptes_aggregation.batch_aggregation.Project")
+    @patch("panoptes_aggregation.batch_aggregation.Panoptes.connect")
+    def test_save_exports(self, mock_client, mock_project, mock_workflow, mock_mkdir):
+        # Test that Panoptes calls are made and files are saved
+        csv_dict = {'media': [{'src': 'http://zooniverse.org/123.csv'}]}
+        mock_project.return_value.describe_export.return_value = csv_dict
+        mock_workflow.return_value.describe_export.return_value = csv_dict
+        ba = batch_agg.BatchAggregator(1, 10, 100)
+        batch_agg.BatchAggregator._download_export = MagicMock(side_effect=['./cls_export.csv', './wf_export.csv'])
+        mock_uuidgen = MagicMock(side_effect=ba._generate_uuid())
+        ba._generate_uuid = mock_uuidgen
+        expected_response = {'classifications': 'tmp/10/10_cls_export.csv', 'workflows': 'tmp/10/10_workflow_export.csv'}
+
+        response = ba.save_exports()
+
+        # Why do these mocked methods called in __init__ not get counted as called?
+        # They are def getting called as the attributes are set
+        # mock_uuidgen.assert_called_once()
```

Comment:
Isn't it enough to test that `ba.id is not None` to ensure the method has run? (i.e. test for the desired result rather than the specific way of getting that result).

Reply:
It probably is, yeah. That'll work for UUID and I'll just remove the client check, but I must be mocking something wrong to see `ba.id` despite the assert failing.